flax/linen/module.py

```diff
@@ -897,6 +910,20 @@ def __getattr__(self, name: str) -> Any:
           'are only accessible from inside \'init\' or \'apply\'.')
       raise AttributeError(msg)
 
+  def __getattribute__(self, name):
+    """Call setup() before accessing any submodule attributes."""
+    # NB: all code here is very "hot" and will be run very frequently.
+    if (object.__getattribute__(self, 'scope') is not None
+        and '_submodule_dataclass_fields' in object.__getattribute__(self, '__dict__')
+        and name in object.__getattribute__(self, '_submodule_dataclass_fields')
+        and name not in self._state.registered_submodules):
+      # object.__getattribute__(self, '_try_setup')()
+      value = object.__getattribute__(self, name)
+      object.__getattribute__(self, '_register_submodules')(name, value)
```

Comment:
I do worry about adding a codepath outside of try_setup here for doing "setup" stuff - just in terms of future readability.

Reply:
Maybe we can just call `_try_setup`, it would be better for determinism.